osgar/lib/quaternion.py

```diff
@@ -36,9 +36,16 @@ def rotate_vector(vector, quaternion):
     return multiply(part1, con)[:-1]
 
 def euler_zyx(quaternion):
+    # https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+    # assumption that quaternion is normalized
     x0, y0, z0, w0 = quaternion
+    sqr_size = x0*x0 + y0*y0 + z0*z0 + w0*w0
+    if abs(sqr_size - 1.0) > 0.00001:
+        k = math.sqrt(sqr_size)
+        x0, y0, z0, w0 = x0/k, y0/k, z0/k, w0/k
```

Comment:
the above comment says the assumption is the quaternion is normalized; first, is that a good assumption, should we protect ourselves from an unexpected input? Second, this block also looks like some kind of normalization, how does it mesh together?

Reply:
well, I added assert there but that would not help you much now :( ... I can add comment that this is the extra normalization if assumption is not satisfied, ok